Optimize module path finding

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52990 Optimize module path finding**

This PR changes module path finding from O(N^2) to O(1)

Differential Revision: [D26779399](https://our.internmc.facebook.com/intern/diff/D26779399)

Commit:
Optimize module path finding

[ghstack-poisoned]

jamesr66a reviewed (approved):
This seems fine to land. We should be on the lookout for if users get confused in the future since we have a) this state now cached in the tracer but b) the underlying Module hierarchy can be modified. Don't need to address that in this PR though

Commit:
Update on "Optimize module path finding"


This PR changes module path finding from O(N^2) to O(1)

[ghstack-poisoned]

jamesr66a commented on torch/fx/symbolic_trace.py:
```diff
@@ -295,6 +299,7 @@ def trace(self, root: Union[torch.nn.Module, Callable], concrete_args: Optional[
         if isinstance(root, torch.nn.Module):
             self.root = root
             fn = type(root).forward
+            self.submodule_paths = {mod: name for name, mod in root.named_modules()}
```
Maybe also null this out when we exit `trace()`? Not sure if it can happen but e.g. if someone uses the same `Tracer` instance to first trace a module but then trace a free function, the free function trace would happen with a `submodule_paths` instance that doesn't belong to it

Commit:
Update on "Optimize module path finding"


This PR changes module path finding from O(N^2) to O(1)

Differential Revision: [D26779399](https://our.internmc.facebook.com/intern/diff/D26779399)

[ghstack-poisoned]
